src/Parser/Desugar.ml

```diff
@@ -500,12 +508,22 @@ and tr_expr (e : Raw.expr) =
     make (EMatch(tr_expr e, [make cl1; make cl2]))
   | ESelect(path, e) ->
     make (EDefs([ make (DOpen(false, path)) ], tr_expr e))
-  | EBOp(exp1,op,exp2) -> 
-    let e1 = tr_expr exp1 and e2 = tr_expr exp2 in 
-    make (EApp(
-      { pos  = Position.join e1.pos op.pos; 
-        data = EApp(tr_bop_to_expr op, e1)
-      }, e2))
+  | EBOp(exp1,op,exp2) ->
+    let with_nowhere data = { pos = Position.nowhere; data = data} in
+    begin match op.data with
+    | "&&" ->
+      tr_expr (make (Raw.EIf(exp1, exp2, with_nowhere (Raw.ECtor("False")))))
```

Comment:
While for correct programs it works well, the error messages for wrong programs should be improved. First of all, a type annotation for `exp1` would help.
```
$ cat test1.dbl
let _ = 42 && True
$ dbl test1.dbl
test1.dbl:1:9-18: fatal error: This pattern matches values of type Int, which is not an ADT
```
The other problem is that `False` is located nowhere.
```
$ cat test2.dbl
let _ = True && 42
$ dbl test2.dbl
<nowhere>:1:1: error: This expression has type Bool, but an expression was expected of type Int
```
Adding type annotation to `exp2` should help. Alternatively, we could inline desugaring `EIf` to `EMatch` and revers the order of clauses.